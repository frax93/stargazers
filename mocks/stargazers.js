export const mockStargazers = [
  {
    "login": "alekseiOdegovC",
    "id": 57904449,
    "node_id": "MDQ6VXNlcjU3OTA0NDQ5",
    "avatar_url": "https://avatars.githubusercontent.com/u/57904449?v=4",
    "gravatar_id": "",
    "url": "https://api.github.com/users/alekseiOdegovC",
    "html_url": "https://github.com/alekseiOdegovC",
    "followers_url": "https://api.github.com/users/alekseiOdegovC/followers",
    "following_url": "https://api.github.com/users/alekseiOdegovC/following{/other_user}",
    "gists_url": "https://api.github.com/users/alekseiOdegovC/gists{/gist_id}",
    "starred_url": "https://api.github.com/users/alekseiOdegovC/starred{/owner}{/repo}",
    "subscriptions_url": "https://api.github.com/users/alekseiOdegovC/subscriptions",
    "organizations_url": "https://api.github.com/users/alekseiOdegovC/orgs",
    "repos_url": "https://api.github.com/users/alekseiOdegovC/repos",
    "events_url": "https://api.github.com/users/alekseiOdegovC/events{/privacy}",
    "received_events_url": "https://api.github.com/users/alekseiOdegovC/received_events",
    "type": "User",
    "site_admin": false
  },
  {
    "login": "prowlee",
    "id": 35212179,
    "node_id": "MDQ6VXNlcjM1MjEyMTc5",
    "avatar_url": "https://avatars.githubusercontent.com/u/35212179?v=4",
    "gravatar_id": "",
    "url": "https://api.github.com/users/prowlee",
    "html_url": "https://github.com/prowlee",
    "followers_url": "https://api.github.com/users/prowlee/followers",
    "following_url": "https://api.github.com/users/prowlee/following{/other_user}",
    "gists_url": "https://api.github.com/users/prowlee/gists{/gist_id}",
    "starred_url": "https://api.github.com/users/prowlee/starred{/owner}{/repo}",
    "subscriptions_url": "https://api.github.com/users/prowlee/subscriptions",
    "organizations_url": "https://api.github.com/users/prowlee/orgs",
    "repos_url": "https://api.github.com/users/prowlee/repos",
    "events_url": "https://api.github.com/users/prowlee/events{/privacy}",
    "received_events_url": "https://api.github.com/users/prowlee/received_events",
    "type": "User",
    "site_admin": false
  }
];
