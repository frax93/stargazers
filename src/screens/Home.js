import { View, StyleSheet, Alert } from 'react-native';
import { useState, useCallback } from 'react';
import { useFormik } from 'formik';
import * as yup from 'yup';
import Spinner from 'react-native-loading-spinner-overlay';
import SearchForm from '../components/SearchForm.js';
import UserList from '../components/UserList.js';

const initialValues = {
  repository: '',
  owner: '',
  stargazersList: null,
};

const validationSchema = yup.object().shape({
  owner: yup.string().required('Field required'),
  repository: yup.string().required('Field required'),
});

export default function HomeScreen() {
  const [isLoading, setIsLoading] = useState(false);

  const onSubmit = useCallback(
    async (values, context) => {
      setIsLoading(true);
      try {
        const res = await fetch(`https://api.github.com/repos/${values.owner}/${values.repository}/stargazers`
        );
        if (res.status !== 200) {
          throw new Error('Error');
        }
        const stargazers = await res.json();
        context.setFieldValue('stargazersList', stargazers);
      } catch (e) {
        console.log(e);
        Alert.alert(
          "Error during search"
        );
        context.setFieldValue('stargazersList', []);
      }
      finally {
        setIsLoading(false);
      }
    }, [setIsLoading]);

  const formikContext = useFormik({
    initialValues,
    onSubmit,
    validationSchema,
  });

  return (
    <View style={styles.container}>
      {isLoading ? <Spinner
        visible={isLoading}
        textContent={'Loading...'}
        textStyle={styles.spinnerTextStyle}
      /> : undefined}
      <SearchForm values={formikContext.values} errors={formikContext.errors} touched={formikContext.touched} handleSubmit={formikContext.handleSubmit} onChange={formikContext.setFieldValue} handleReset={formikContext.handleReset} />
      <UserList data={formikContext.values.stargazersList} />
    </View>
  );
}

const styles = StyleSheet.create({
  spinnerTextStyle: {
    color: '#FFF',
  },
  container: {
    flex: 1,
    backgroundColor: '#ecf0f1',
    padding: 8,
  },
  title: {
    margin: 24,
    fontSize: 24,
    fontWeight: 'bold',
    textAlign: 'center',
  },
});
