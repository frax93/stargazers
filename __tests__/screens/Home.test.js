import Home from 'screens/Home';
import { render, fireEvent, waitFor } from '@testing-library/react-native';
import { mockStargazers } from '../../mocks/stargazers';

const setup = () => {
  const wrapper = render(<Home />);
  return wrapper;
};


global.fetch = jest.fn((url) => {
  return new Promise((resolve, reject) => {
    setTimeout(() => {
      if(url.includes('burzaszsolt')) {

        resolve({
          json: () => Promise.resolve(mockStargazers),
          status: 200
        });
        return;
      }

      reject({ status: 404 });
    }, 3000);

  });
}
);

describe('Testing home screen', () => {
  it("renders without errors", () => {
    const homeScreenComponent = setup();

    const componentToJSON = homeScreenComponent.toJSON();

    expect(componentToJSON).toMatchSnapshot();
  });

  it("does not show validation errors", async () => {
    const { getByTestId, queryAllByText } = setup();


    const repositoryInput = getByTestId('repository');
    const ownerInput = getByTestId('owner');
    const submitButton = getByTestId('submit');

    await waitFor(() => {
      fireEvent.changeText(repositoryInput, 'repository');
    });

    await waitFor(() => {
      fireEvent.changeText(ownerInput, 'burzaszsolt');
    });

    await waitFor(() => {
      fireEvent.press(submitButton);
    });

    await waitFor(() => {
      expect(queryAllByText('Field required')).toHaveLength(0);
    }, { timeout: 4000 });

  });

  it("is loading spinner is present", async () => {
    const { getByTestId, findByText } = setup();


    const repositoryInput = getByTestId('repository');
    const ownerInput = getByTestId('owner');
    const submitButton = getByTestId('submit');

    await waitFor(() => {
      fireEvent.changeText(repositoryInput, 'repository');
    });

    await waitFor(() => {
      fireEvent.changeText(ownerInput, 'burzaszsolt');
    });

    await waitFor(() => {
      fireEvent.press(submitButton);
    });

    expect((await (await findByText('Loading...', {}, { timeout: 100 })).children[0].children[0])).toBe('Loading...');

  });

  it("fetch is ok", async () => {
    const { getByTestId, getAllByTestId } = setup();


    const repositoryInput = getByTestId('repository');
    const ownerInput = getByTestId('owner');
    const submitButton = getByTestId('submit');

    await waitFor(() => {
      fireEvent.changeText(repositoryInput, 'repository');
    });

    await waitFor(() => {
      fireEvent.changeText(ownerInput, 'burzaszsolt');
    });

    await waitFor(() => {
      fireEvent.press(submitButton);
    });


    await waitFor(() => {
      const imageItems = getAllByTestId('image-item');
      const avatarItems = getAllByTestId('avatar-item');

      const [firstImageItem] = imageItems || [];
      const [firstAvatarItem] = avatarItems || [];

      expect(firstImageItem.props.source).toHaveProperty('uri', 'https://avatars.githubusercontent.com/u/57904449?v=4');

      expect(firstAvatarItem.children[0]).toBe('alekseiOdegovC');
    }, { timeout: 4000 });

  });

  it("fetch is ko", async () => {
    const { getByTestId, queryAllByText } = setup();


    const repositoryInput = getByTestId('repository');
    const ownerInput = getByTestId('owner');
    const submitButton = getByTestId('submit');

    await waitFor(() => {
      fireEvent.changeText(repositoryInput, 'repository');
    });

    await waitFor(() => {
      fireEvent.changeText(ownerInput, 'owner');
    });

    await waitFor(() => {
      fireEvent.press(submitButton);
    });


    await waitFor(() => {
      const noTextElement = getByTestId('no-element');

      expect(noTextElement.children[0]).toBe('No element found');
      expect(queryAllByText('Error during search')).not.toBe(null);
    }, { timeout: 4000 });

  });


  it("Reset search form", async () => {
    const { getByTestId } = setup();

    const repositoryInput = getByTestId('repository');
    const ownerInput = getByTestId('owner');

    await waitFor(() => {
      fireEvent.changeText(repositoryInput, 'repository');
    });

    await waitFor(() => {
      fireEvent.changeText(ownerInput, 'owner');
    });

    await waitFor(() => {
      expect(getByTestId('repository').props.value).toBe('repository');
      expect(getByTestId('owner').props.value).toBe('owner');
    });

    const resetButton = getByTestId('reset');

    await waitFor(() => {
      fireEvent.press(resetButton);
    });

    await waitFor(() => {
      expect(getByTestId('repository').props.value).toBe('');
      expect(getByTestId('owner').props.value).toBe('');
    });

  });

  it("Show validation search errors", async () => {
    const { getByTestId, queryAllByText } = setup();

    const submitButton = getByTestId('submit');

    await waitFor(() => {
      fireEvent.press(submitButton);
    });

    await waitFor(() => {
      expect(queryAllByText('Field required')).toHaveLength(2);
    });

  });
});

