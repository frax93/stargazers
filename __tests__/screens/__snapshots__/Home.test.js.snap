// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing home screen renders without errors 1`] = `
<View
  style={
    {
      "backgroundColor": "#ecf0f1",
      "flex": 1,
      "padding": 8,
    }
  }
>
  <View>
    <TextInput
      nativeID="repository"
      onChangeText={[Function]}
      placeholder="Enter repository"
      style={
        {
          "backgroundColor": "#fff",
          "borderColor": "#ddd",
          "borderWidth": 1,
          "height": 50,
          "paddingHorizontal": 8,
          "width": "100%",
        }
      }
      testID="repository"
      value=""
    />
    <TextInput
      nativeID="owner"
      onChangeText={[Function]}
      placeholder="Enter owner"
      style={
        {
          "backgroundColor": "#fff",
          "borderColor": "#ddd",
          "borderWidth": 1,
          "height": 50,
          "paddingHorizontal": 8,
          "width": "100%",
        }
      }
      testID="owner"
      value=""
    />
    <View
      collapsable={false}
      nativeID="submit"
      style={
        {
          "backgroundColor": "black",
          "borderColor": "transparent",
          "borderRadius": 4,
          "borderStyle": "solid",
          "borderWidth": 0,
          "elevation": 2,
          "marginTop": 16,
          "minWidth": 64,
          "shadowColor": "#000000",
          "shadowOffset": {
            "height": 0.75,
            "width": 0,
          },
          "shadowOpacity": 0.24,
          "shadowRadius": 1.5,
        }
      }
    >
      <View
        accessibilityRole="button"
        accessibilityState={
          {
            "disabled": false,
          }
        }
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          [
            {
              "overflow": "hidden",
            },
            {
              "borderRadius": 4,
            },
          ]
        }
        testID="submit"
      >
        <View
          style={
            [
              {
                "alignItems": "center",
                "flexDirection": "row",
                "justifyContent": "center",
              },
              undefined,
            ]
          }
        >
          <Text
            numberOfLines={1}
            selectable={false}
            style={
              [
                {
                  "color": "#000000",
                  "fontFamily": "System",
                  "fontWeight": "400",
                },
                {
                  "textAlign": "left",
                },
                [
                  {
                    "letterSpacing": 1,
                    "marginHorizontal": 16,
                    "marginVertical": 9,
                    "textAlign": "center",
                  },
                  undefined,
                  {
                    "textTransform": "uppercase",
                  },
                  {
                    "color": "#ffffff",
                    "fontFamily": "System",
                    "fontWeight": "500",
                  },
                  {
                    "fontFamily": "System",
                    "fontWeight": "500",
                  },
                  undefined,
                ],
              ]
            }
          >
            Submit
          </Text>
        </View>
      </View>
    </View>
    <View
      collapsable={false}
      nativeID="reset"
      style={
        {
          "backgroundColor": "transparent",
          "borderColor": "rgba(0, 0, 0, 0.29)",
          "borderRadius": 4,
          "borderStyle": "solid",
          "borderWidth": 0.5,
          "elevation": 0,
          "marginTop": 16,
          "minWidth": 64,
        }
      }
    >
      <View
        accessibilityRole="button"
        accessibilityState={
          {
            "disabled": false,
          }
        }
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          [
            {
              "overflow": "hidden",
            },
            {
              "borderRadius": 4,
            },
          ]
        }
        testID="reset"
      >
        <View
          style={
            [
              {
                "alignItems": "center",
                "flexDirection": "row",
                "justifyContent": "center",
              },
              undefined,
            ]
          }
        >
          <Text
            numberOfLines={1}
            selectable={false}
            style={
              [
                {
                  "color": "#000000",
                  "fontFamily": "System",
                  "fontWeight": "400",
                },
                {
                  "textAlign": "left",
                },
                [
                  {
                    "letterSpacing": 1,
                    "marginHorizontal": 16,
                    "marginVertical": 9,
                    "textAlign": "center",
                  },
                  undefined,
                  {
                    "textTransform": "uppercase",
                  },
                  {
                    "color": "black",
                    "fontFamily": "System",
                    "fontWeight": "500",
                  },
                  {
                    "fontFamily": "System",
                    "fontWeight": "500",
                  },
                  undefined,
                ],
              ]
            }
          >
            Reset
          </Text>
        </View>
      </View>
    </View>
  </View>
  <View />
</View>
`;
